query/src/sql/optimizer/rule/rule_implement_hash_join.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_exception::Result;
+
+use crate::sql::optimizer::rule::transform_state::TransformState;
+use crate::sql::optimizer::rule::Rule;
+use crate::sql::optimizer::rule::RuleID;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::LogicalInnerJoin;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::PhysicalHashJoin;
+use crate::sql::plans::PlanType;
+
+pub struct RuleImplementHashJoin {
+    id: RuleID,
+    pattern: SExpr,
+}
+
+impl RuleImplementHashJoin {
+    pub fn create() -> Self {
+        RuleImplementHashJoin {
+            id: RuleID::ImplementHashJoin,
+            pattern: SExpr::create_binary(
+                PatternPlan {
+                    plan_type: PlanType::LogicalInnerJoin,
+                }
+                .into(),
+                SExpr::create_leaf(
+                    PatternPlan {
+                        plan_type: PlanType::Pattern,
+                    }
+                    .into(),
+                ),
+                SExpr::create_leaf(
+                    PatternPlan {
+                        plan_type: PlanType::Pattern,
+                    }
+                    .into(),
+                ),
+            ),
+        }
+    }
+}
+
+impl Rule for RuleImplementHashJoin {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, expression: &SExpr, state: &mut TransformState) -> Result<()> {
+        let plan = expression.plan();
+        let logical_inner_join: LogicalInnerJoin = plan.try_into()?;
+
+        let result = SExpr::create(
+            PhysicalHashJoin {
```

Comment:
In the future, when our statistics module matures, it seems that the build side can be determined based on statistical information

Reply:
Yes. We will always choose right side as build side. The join commutativity rule can help us enumurate the candidates.